primefaces/src/main/java/org/primefaces/model/JpaLazyDataModel.java

```diff
@@ -335,214 +328,132 @@ protected Expression resolveFieldExpression(CriteriaBuilder cb, CriteriaQuery<?>
         }
 
+        @Deprecated
         public Builder<T> rowKey(String rowKey) {
-            this.rowKey = rowKey;
+            model.rowKeyField = rowKey;
             return this;
         }
 
+        @Deprecated
         public Builder<T> rowKey(SingularAttribute<T, ?> rowKeyMetamodel) {
-            this.rowKey = rowKeyMetamodel.getName();
+            model.rowKeyField = rowKeyMetamodel.getName();
+            return this;
+        }
+
+        public Builder<T> rowKeyField(String rowKey) {
+            model.rowKeyField = rowKey;
+            return this;
+        }
+
+        public Builder<T> rowKeyField(SingularAttribute<T, ?> rowKeyMetamodel) {
+            model.rowKeyField = rowKeyMetamodel.getName();
             return this;
         }
 
         public Builder<T> caseSensitive(boolean caseSensitive) {
-            this.caseSensitive = caseSensitive;
+            model.caseSensitive = caseSensitive;
             return this;
         }
 
         public Builder<T> wildcardSupport(boolean wildcardSupport) {
-            this.wildcardSupport = wildcardSupport;
+            model.wildcardSupport = wildcardSupport;
             return this;
         }
 
         public Builder<T> queryEnricher(QueryEnricher<T> queryEnricher) {
-            this.queryEnricher = queryEnricher;
+            model.queryEnricher = queryEnricher;
             return this;
         }
 
         public Builder<T> filterEnricher(FilterEnricher<T> filterEnricher) {
-            this.filterEnricher = filterEnricher;
+            model.filterEnricher = filterEnricher;
             return this;
         }
 
         public Builder<T> sortEnricher(SortEnricher<T> sortEnricher) {
-            this.sortEnricher = sortEnricher;
+            model.sortEnricher = sortEnricher;
+            return this;
+        }
+
+        public Builder<T> rowKeyType(Class<?> rowKeyType) {
+            model.rowKeyType = rowKeyType;
             return this;
         }
 
         public JpaLazyDataModel<T> build() {
-            JpaLazyDataModel<T> model = new JpaLazyDataModel<>(entityClass, entityManager);
-            model.rowKey = rowKey;
-            model.caseSensitive = caseSensitive;
-            model.wildcardSupport = wildcardSupport;
-            model.queryEnricher = queryEnricher;
-            model.filterEnricher = filterEnricher;
-            model.sortEnricher = sortEnricher;
-            model.setRowKeyConverter(rowKeyConverter);
+            Objects.requireNonNull(model.entityClass, "entityClass not set");
+            Objects.requireNonNull(model.entityManager, "entityManager not set");
 
+            if (model.rowKeyProvider == null) {
+                if (model.rowKeyConverter != null) {
+                    model.rowKeyProvider = model::getRowKeyFromConverter;
+                }
+                else {
+                    Objects.requireNonNull(model.rowKeyField, "rowKeyField is mandatory if no rowKeyProvider nor converter is provided");
```

Comment:
actually not true, its only required if you have selection enabled
therefore i made the initilization lazy